easybuild/tools/job/pyslurm_backend.py

```diff
@@ -0,0 +1,146 @@
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+Interface module to SLURM, via PySlurm
+
+:author: Kenneth Hoste (Ghent University)
+"""
+from distutils.version import LooseVersion
+from vsc.utils import fancylogger
+
+from easybuild.tools.build_log import EasyBuildError, print_msg
+from easybuild.tools.config import build_option
+from easybuild.tools.job.backend import JobBackend
+from easybuild.tools.run import run_cmd
+from easybuild.tools.utilities import only_if_module_is_available
+
+
+_log = fancylogger.getLogger('pyslurm_backend', fname=False)
+
+
+try:
+    import pyslurm
+    from pyslurm import version as pyslurm_version
+except ImportError as err:
+    _log.debug("Failed to import pyslurm"
+               " Silently ignoring, this is a real issue only when PySlurm is used as backend for --job")
+
+
+class PySlurm(JobBackend):
+    """
+    Manage SLURM server communication and create `SlurmJob` objects.
+    """
+
+    # only tested with PySLURM 17.x (and newer)
+    REQ_VERSION = '17'
+
+    @only_if_module_is_available('pyslurm', pkgname='pyslurm')
+    def __init__(self, *args, **kwargs):
+        """Constructor."""
+        super(PySlurm, self).__init__(*args, **kwargs)
+
+    # _check_version is called by __init__, so guard it (too) with the decorator
+    @only_if_module_is_available('pyslurm', pkgname='pyslurm')
+    def _check_version(self):
+        """Check whether pbs_python version complies with required version."""
+        version = pyslurm_version()
+        if LooseVersion(version) < LooseVersion(self.REQ_VERSION):
+            raise EasyBuildError("Found pyslurm version %s, but version %s or more recent is required",
+                                 version, self.REQ_VERSION)
+
+    def init(self):
+        """
+        Initialise the job backend.
+
+        Connect to the PBS server & reset list of submitted jobs.
+        """
+        self._submitted = []
+
+    def queue(self, job, dependencies=frozenset()):
+        """
+        Add a job to the queue.
+
+        :param dependencies: jobs on which this job depends.
+        """
+        if dependencies:
+            job.job_specs['dependency'] = 'afterany:%s' % ':'.join(str(d.jobid) for d in dependencies)
```

Comment:
Should probably be "afterok", since it's likely useless to continue if the dependency failed.

Reply:
We also use `afterany` with `pbs_pyton`, so I prefer keeping this consistent...

It's true that the dependent jobs will fail, but at least it'll be clear why (i.e. a missing module), while if they just disappear into thin air, it may not be clear at all...

Reply:
Maybe so, but using afterany consumes resources when it shouldn't. And if the preceding dependency failed you wouldn't care about the next step.

Reply:
What about adding a configuration option for this? Something like `--job-deps-type=afterok`?

Reply:
Yeah, that would work. And it would be backend agnostic if done correctly. (whatever that is in this case)